Fix most ports being built without optimizations (e.g. -O3) on *NIX

lhecker opened the pull request:
## Issue

Using vcpkg I built OpenCV on Linux (Ubuntu 18.10), but I was highly surprised with OpenCV's plain horrible performance for its release builds.

All I did was:
```sh
./vcpkg install opencv
```

After some debugging it turns out that OpenCV was built without **ANY** kind of optimization flags for the release build.

**Same for libjpeg-turbo, libtiff etc. though.** :warning: 

## Fix

**tl;dr:** See down below in section "Commit message"

In your local vcpkg setup you'll find a folder at:
```
./downloads/tools/cmake-*/cmake-*/share/cmake-*/Modules/Compiler
```
This folder contains various compiler-specific definitions for those that CMake supports.

For `gcc` (i.e. `GNU.cmake`) for instance this entails the following very important definition:
```cmake
string(APPEND CMAKE_${lang}_FLAGS_RELEASE_INIT " -O3 -DNDEBUG")
```
But those flags only apply if the user hasn't set the corresponding (e.g.) `CMAKE_${lang}_FLAGS_RELEASE` variable yet, which we did though using e.g. `set(CMAKE_CXX_FLAGS_RELEASE ...)` etc.

The optimal solution would be that we only use the `_INIT` variables from now on like above in our toolchains, but unfortunately this doesn't guarantee that the `VCPKG_CXX_FLAGS_RELEASE` flags take precedence anymore.
(In fact, as can be seen above, CMake appends `-O3 -DNDEBUG` and would thus overwrite whatever you set in `VCPKG_CXX_FLAGS_RELEASE`.)

So as far as I can see, without breaking backwards compatibility, the only solution is the one I'm proposing in this PR. :slightly_smiling_face:

P.S.: I know this might sound rude, but I _am_ seriously whondering why no one noticed the horrendous performance of libraries built with vcpkg on *NIX before... :thinking: 

## Commit message
```
Don't set CMAKE_*_FLAGS* unless necessary

If a CMAKE_*_FLAGS* variable is set, cmake's compiler-specific
toolchain settings will not be applied.
The way the previous scripts/toolchain/*.cmake code was written
these flags where set to empty or whitespace-only strings and
caused cmake to not apply any compiler-specific settings,
like e.g. the common "-O3 -DNDEBUG" for release builds.
```

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=4816) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=4816) before we can accept your contribution.<br/><br/>:x: lhecker [sign now](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=4816)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vcpkg?pullRequest=4816) it.</sub>

ras0219-msft commented:
Thanks a bunch for the PR and noticing this issue!

We very recently also found this due to a build break in release by not setting `-DNDEBUG` (https://github.com/Microsoft/vcpkg/issues/4731, though I failed to record the specific details).

I completely concur with your analysis about this being the only solution that doesn't change behavior of `VCPKG_CXX_FLAGS` (save the fact that it would "overwrite" `-fPIC`, but see below).

However, I think in this case, it is desirable to have `VCPKG_CXX_FLAGS` apply to `CMAKE_CXX_FLAGS_INIT` instead of `CMAKE_CXX_FLAGS`. The intention is really to have a mechanism similar to `$CXXFLAGS`[1] where you can just add a small tweak to the build; for users who need to totally replace everything, we'd recommend they write a proper full toolchain file.

I've been testing `dev/roschuma/4731`[2], does that look like it would also solve the issue to you? (Edit: aha, I accidentally skipped freebsd!)

[1] In the core CMake files, `$CXXFLAGS` is applied to `XYZ_INIT` as well https://github.com/Kitware/CMake/blob/master/Modules/CMakeCXXInformation.cmake#L200
[2] https://github.com/Microsoft/vcpkg/commit/cc20d43686217009f94c60dd8c54290a5e76aa61

lhecker committed:
Use the _INIT variant of CMAKE_*_FLAGS*

If a CMAKE_*_FLAGS* variable is set, cmake's compiler-specific
toolchain settings will not be applied.
The way the previous scripts/toolchain/*.cmake code was written
these flags where set to empty or whitespace-only strings and
caused cmake to not apply any compiler-specific settings,
like e.g. the common "-O3 -DNDEBUG" for release builds.

lhecker commented:
@ras0219-msft Yes that branch would work as well. 🙂

Since you said that you're fine with the slight behavior change of the `VCPKG_*_FLAGS*` variables I've taken the liberty to rebase my PR from e429aa0 to 390a517, which contains my original attempt.
Incidentially this is then almost identical to your branch.
So please feel free to close this PR and merge your changes if you want to. 🙂

lhecker commented:
Fixed in 4d1fd42fca46f4c4dfe35011c8c96ff941db3ca6 :point_right: Closing.
cc: @ras0219-msft
